src/Microsoft.AspNetCore.SignalR.Core/HubGroupList.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.SignalR
+{
+    public class HubGroupList : IReadOnlyCollection<ConcurrentDictionary<string, HubConnectionContext>>
+    {
+        private readonly ConcurrentDictionary<string, GroupConnectionList> _groups =
+            new ConcurrentDictionary<string, GroupConnectionList>();
+
+        private static readonly GroupConnectionList EmptyGroupConnectionList = new GroupConnectionList();
+
+        public ConcurrentDictionary<string, HubConnectionContext> this[string groupName]
+        {
+            get
+            {
+                _groups.TryGetValue(groupName, out var group);
+                return group;
+            }
+        }
+
+        public void Add(HubConnectionContext connection, string groupName)
+        {
+            CreateOrUpdateGroupWithConnection(groupName, connection);
+        }
+
+        public void Remove(string connectionId, string groupName)
+        {
+            if (!_groups.TryGetValue(groupName, out var connections)) return;
+            ICollection<KeyValuePair<string, GroupConnectionList>> col = _groups;
+            if (!connections.TryRemove(connectionId, out var _) || !connections.IsEmpty) return;
+            var groupToRemove =
+                new KeyValuePair<string, GroupConnectionList>(groupName, EmptyGroupConnectionList);
+            col.Remove(groupToRemove);
```

Comment:
Either have `col` defined next to this or just cast `_groups` inline

Reply:
We should have a comment here explaining exactly what is happening.

Reply:
Comment here or in code?

Reply:
Code, we don't want to see this code in a couple months and have no idea what its doing :D